src/bucket/BucketManagerImpl.cpp

```diff
@@ -94,7 +94,7 @@ BucketManagerImpl::BucketManagerImpl(Application& app)
     , mBucketSnapMerge(app.getMetrics().NewTimer({"bucket", "snap", "merge"}))
     , mSharedBucketsSize(
           app.getMetrics().NewCounter({"bucket", "memory", "shared"}))
-
+    , mMergeFutureCache(1024)
```

Comment:
why 1024? isn't the worst case around 10 merges? Given that you're using `weak_ptr`, we should never go beyond that (and 1024 is a huge number when translated into buckets).

In general, I don't know if a cache is the right data structure: what we really want is to remove merges that are not in use anymore (`weak_ptr` expired).

Reply:
1024 is a little arbitrary; there are enough different paths that intersect here that I figured it wouldn't hurt to give it some wiggle room.

I think the _worst_ case is when we're doing rapid replay like during catchup-with-publish: we kick off 10 merges passing a publish checkpoint, serialize an HAS (forgetting those 10), then continue setting off merges at maximum replay-speed. If the HAS is not deserialized soon enough (and it might not -- there might be a backlog of publishing) then the long running merges (say L8->L9) it was supposed to re-attach to will be purged from the cache.

Another way to look at it is that the cache should really only bother with longer-lived merges. So we could theoretically make it smaller and exclude merges younger than L4 or something? But I figured this is harmless as-is.

I don't know why a cache would be a bad data structure. A `weak_ptr` is a small thing to keep around when its corresponding `shared_ptr` is dead: just a control block.

(Oh! I just looked and [according to this](https://www.bfilipek.com/2017/12/weakptr-memory.html) it seems like `shared_ptr` will combine the control block with the pointee allocation when you use `make_shared` rather than constructing from a `new` ptr, so this might retain an entire destructed-but-allocated `Bucket` structure for the `weak_ptr`. That's still not .. especially big though. A destructed `std::string` is 32 bytes, a `Hash` is 20 bytes, and a `size_t` is 8 bytes. So a `Bucket` in that state will be 60 bytes + the control block which I think is 16 bytes? 1024 of these is ~77k.)

Reply:
Note to self: followed-up in chat and the concern here isn't the size, it's that a cache might lose a live merge for no reason (regardless of size) whereas there are deterministic lifecycle edges for a FutureBucket at which a running merge can be registered and de-registered precisely.